prow/resultstore/payload.go

```diff
@@ -151,37 +160,68 @@ func (p *Payload) invocationProperties() []*resultstore.Property {
 			Value: p.Job.Status.URL,
 		},
 	}
-	if p.Job.Spec.Refs != nil {
-		ps = append(ps, &resultstore.Property{
-			Key:   "Repo",
-			Value: p.Job.Spec.Refs.RepoLink,
-		})
-	}
-	if p.Started != nil {
-		ps = append(ps, &resultstore.Property{
-			Key:   "Commit",
-			Value: p.Started.RepoCommit,
-		})
-		for _, branch := range p.Started.Repos {
-			ps = append(ps, &resultstore.Property{
-				Key:   "Branch",
-				Value: branch,
-			})
-		}
+	ps = append(ps, p.podSpecProperties()...)
+	ps = append(ps, p.startedProperties()...)
+	return ps
+}
+
+func (p *Payload) podSpecProperties() []*resultstore.Property {
+	var ps []*resultstore.Property
+	if p.Job.Spec.PodSpec == nil {
```

Comment:
Earlier up you did `p.Job != nil` so this check to go all the way down to the last member feels like it can blow up at runtime if `p.Job` is nil... WDYT?

Reply:
This was discussed briefly in the original PR. These helper functions are intended to simplify the Invocation() function which guards p.Job.

I've striven to avoid redundant checks and test the intended entry points against nil pointers; I can make this more defensive and (re-)check at every function, but it diminishes the simplification intended by using helper functions. I'm still relatively new to Prow, and don't know the balance sought.

Factoring this into it's own package seems overkill, but I was considering capitalizing methods expected to be called from outside the class (and therefore must be defensive). Do you think this would improve readability?

Reply:
>  I can make this more defensive and (re-)check at every function, but it diminishes the simplification intended by using helper functions

I see two options: either (1) check at every function, or (2) not do the check once and also guarantee that it won't result in a nil pointer exception downstream. If (2) is possible, we can do that (basically do some validation/defaulting and then use a new type that is guaranteed to never have problematic nils) instead of checking all the time. You could do (1) if it's not too onerous, and move to (2) later. Or just do (2) now. I suggest doing whatever is easiest.

> I'm still relatively new to Prow, and don't know the balance sought.

In Prow I've seen (2) being done in the case of job defaulting (we don't use raw job configs but default them first and then use the defaulted versions).

> Factoring this into it's own package seems overkill, but I was considering capitalizing methods expected to be called from outside the class (and therefore must be defensive). Do you think this would improve readability?

I don't understand what your concern is around readability. Either we can have the helper functions as they are, or put them in a separate package. Either way we have the same amount of code... unless I'm missing something. If you're concerned about another layer of indirection (having to use a new package) hurting readability, I agree. But here I care much more about avoiding runtime nil panics than readability.

Reply:
> > Factoring this into it's own package seems overkill, but I was considering capitalizing methods expected to be called from outside the class (and therefore must be defensive). Do you think this would improve readability?
> 
> I don't understand what your concern is around readability.

To clarify, my concern is that having all "private" methods makes it hard for the reader to know which entry points must be guarded.

Had I made payload a separate package, methods would be explicitly Public and private. It would be easy to see that Invocation() is a public entry point and the helpers are all private. In my experience, private functions are allowed assumptions not made for public ones. However looking more at this I believe the best approach is to make the helpers pure functions as I floated in another comment.

Reply:
I've refactored to address all possible nil pointer conditions now. It is a new commit since merge conflicts would be substantial.